feat(frontend): add view more popup modal for essay questions

ericsen-tsai opened the pull request:
## Summary

This PR adds a new modal dialog that displays all essay questions for a post, allowing users to view and interact with answers in a dedicated popup interface. The modal includes features such as like/unlike functionality with optimistic UI updates, infinite scrolling for answers, and responsive design for mobile and desktop.

## Changes

### Backend (CMS)
- Added new GraphQL query `getMemberEssayAnswersHasLiked` to check if a member has liked specific essay answers
- Implemented member-owned private field access control for sensitive fields (`contactEmail`, `twreporter_user_id`, `showBaodaozai`, `essayQuestionCount`, `createdAt`, `updatedAt`)
- Updated Member list access control to allow all roles to query while protecting private fields
- Added `memberOwnedPrivateFieldQueryAccess` utility function for field-level access control

### Frontend API Layer
- Added API functions for post essay answer likes (create/delete mutations)
- Added API function `getPostEssayQuestionsByPostSlug` to fetch essay questions by post slug
- Added API function `getMemberEssayAnswersHasLiked` to check like status for multiple answers
- Created GraphQL queries for essay question answers with pagination support
- Updated existing GraphQL queries to include additional fields (`contactEmail`, `twreporter_user_id`, `post.slug`)

### Frontend React Query Hooks
- Added `useCreatePostEssayAnswerLikeMutation` and `useDeletePostEssayAnswerLikeMutation` hooks for like operations
- Added `usePostEssayQuestionEssayAnswersInfinityQuery` hook for infinite scrolling of answers
- Added `useGetMemberEssayAnswersHasLikedQuery` hook to fetch like status
- Added `usePostEssayQuestionsByPostSlugQuery` hook to fetch questions by post slug
- Renamed `useGetPostsEssayAnswersWithLikesInfinityQuery` to `usePostsEssayAnswersWithLikesInfinityQuery`

### Frontend UI Components
- Created new `Dialog` component based on Radix UI with responsive design (mobile full-screen, tablet/desktop centered modal)
- Added new icons: `ArticleShortcutIconLarge`, `XIcon`, arrow icons (`ArrowLeftIcon`, `ArrowRightIcon`)
- Created `PostEssayQuestionsModal` component - main modal that displays post header and essay questions
- Created `ModalQuestionCard` component - accordion-style card for each question with infinite scrolling answers
- Created `ModalAnswerItem` component - individual answer item with like functionality
- Created `ModalAnswerItemSkeleton` component - loading skeleton for answers
- Updated existing answer card components to support like functionality and modal trigger

### Frontend Features
- Implemented optimistic UI updates for like/unlike actions using `useOptimisticLikeAnswer` hook
- Added infinite scrolling with intersection observer for tablet/desktop
- Added "顯示更多" (Show More) button for mobile pagination
- Implemented scroll-based title truncation for better UX when scrolling
- Added article shortcut link in modal header to navigate to full article
- Integrated modal with existing Idea Hub modules (`all-answers`, `latest-answers`)

### Dependencies
- Added `@radix-ui/react-dialog` package for dialog component

## Additional Notes

- The modal uses a responsive design: full-screen on mobile, centered modal on tablet/desktop
- Like functionality includes optimistic updates with automatic rollback on error
- Access control changes ensure private member fields are only visible to admins or the member themselves
- The modal supports multiple questions displayed in an accordion layout on mobile and side-by-side cards on tablet/desktop
- All GraphQL queries include proper error handling and authentication checks
- This PR is based on #850 


## Screenshot(s)

<img width="1549" height="1008" alt="image" src="https://github.com/user-attachments/assets/1f64e46d-18ee-4216-8ee7-5f5a17f28e50" />


## Reference(s)

- [Notion Page](https://www.notion.so/twreporter/popup-UI-Integration-2b78dbdca932803188b2c8e9a5dc9abe?v=1588dbdca93281dfa43c000c8415265e&source=copy_link)
- [Figma Design](https://www.figma.com/design/jKXceVhx5seAXU9A93znbK/%E9%96%8B%E6%94%BE%E9%A1%8C%E9%9B%86%E5%90%88%E9%A0%81?node-id=173-198753&p=f&m=dev)

ericsen-tsai committed:
feat(cms, frontend): add essay answer liking functionality

- Introduced GraphQL query `getMemberEssayAnswersHasLiked` to check if a member has liked specific essay answers.
- Implemented corresponding API functions and React Query hooks for fetching liked status.
- Created mutations for liking and unliking essay answers.

ericsen-tsai committed:
chore(frontend): add radix-ui/react-dialog dependency

ericsen-tsai committed:
refactor(frontend): streamline essay question answer fetching and update GraphQL queries

- Removed unused parameters from the `getPostEssayQuestionEssayAnswers` function and updated its API call.
- Enhanced GraphQL queries to include `answerSkip` for pagination and added `slug` to the post data structure.
- Refactored React Query hooks for improved clarity and consistency in fetching essay answers.

ericsen-tsai committed:
feat(frontend): implement dialog component with Radix UI

- Added a new Dialog component along with its subcomponents (DialogTrigger, DialogPortal, DialogClose, DialogOverlay, DialogContent, DialogHeader, DialogFooter, DialogTitle, DialogDescription) to enhance modal functionality.
- Integrated styling and animations for different states of the dialog.
- Included a close button with an XIcon for user accessibility.

ericsen-tsai committed:
feat(frontend): add post essay questions modal and answer components

- Introduced a new modal component for displaying post essay questions, including a dialog for user interaction.
- Implemented modal question cards to present individual questions and their corresponding answers.
- Added functionality for liking answers with optimistic updates to enhance user experience.
- Included skeleton loading states for answers to improve perceived performance during data fetching.
- Refactored existing hooks and components to support the new modal structure and answer interactions.

ericsen-tsai committed:
feat(frontend): enhance Idea Hub with modal and like count display

- Added a modal for posting essay questions, allowing users to view and interact with selected posts.
- Integrated like count display logic across various components using a new utility function for consistency.
- Updated AllAnswers and LatestAnswers components to support modal interactions and improved user feedback with toast notifications.
- Refactored AnswerCard and QuestionAnswerCard to handle modal opening and display likes count effectively.

ericsen-tsai committed:
feat(frontend): update load more functionality in modal question card

- Enhanced the load more feature in the ModalQuestionCard to conditionally render based on device type (tablet vs. non-tablet).
- Updated the IntersectionObserver logic to include a check for the isTablet state, ensuring proper loading behavior across devices.
- Added a button for non-tablet users to fetch more questions, improving accessibility and user experience.

ericsen-tsai committed:
feat(cms): implement member-owned private field access control

- Added `memberOwnedPrivateFieldQueryAccess` utility to restrict access to specific fields based on user role and ownership.

ericsen-tsai committed:
feat(frontend): add corner out icon to post answer card to open post essay questions modal

ericsen-tsai committed:
fix(frontend): copilot comments

- Removed `memberId` parameter from `getMemberPostsWithAnswers` and `getMemberEssayAnswersHasLiked` queries, utilizing session data instead.
- Updated GraphQL queries and API functions to reflect the changes in parameters.
- Enhanced error messages for unauthorized access checks.
- Improved type definitions and added necessary imports for better clarity and maintainability.

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/all-answers/index.tsx:
```diff
@@ -0,0 +1,217 @@
+      const titleRect = titleRef.current.getBoundingClientRect()
+      const threshold = 120
+      if (titleRect.top < threshold) {
+        console.log('trigger!')
```
fixed in [fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -578,6 +578,134 @@ export const extendGraphqlSchema = graphql.extend(() => {
+              )
+              return {
+                answerId: id,
+                hasLiked: answer ? true : false,
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -578,6 +578,134 @@ export const extendGraphqlSchema = graphql.extend(() => {
+          if (isUnauthorized || isForbidden) {
+            const errorMessage = isUnauthorized
+              ? 'Unauthorized to get member essay answers has like'
+              : 'Forbidden to get member essay answers has like'
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/post-essay-questions-modal/modal-question-card.tsx:
```diff
@@ -0,0 +1,227 @@
+                <Divider className="my-4" />
+                <ModalAnswerItemSkeleton />
+              </>
+            )}
+            {!isTablet && hasNextPage && !isFetchingNextPage && (
+              <button
+                onClick={() => fetchNextPage()}
+                className="mt-2 flex w-full cursor-pointer items-center justify-center px-5 py-1 prose-p1 text-neutral-600 hover:text-red-400 active:text-red-500 disabled:cursor-not-allowed disabled:text-neutral-400"
+                disabled={isFetchingNextPage}
+              >
+                顯示更多
+              </button>
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/api/extended.ts:
```diff
@@ -62,3 +65,23 @@ export const getMemberPostsWithAnswers = async (
+  }>({
+    query: GET_MEMBER_ESSAY_ANSWERS_HAS_LIKED_GQL,
+    variables,
+  })
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/post-essay-questions-modal/index.tsx:
```diff
@@ -0,0 +1,229 @@
+        showCloseButton={false}
+      >
+        <DialogHeader>
+          <DialogTitle className="sr-only">Post Essay Questions</DialogTitle>
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/post-essay-questions-modal/modal-answer-item.tsx:
```diff
@@ -0,0 +1,112 @@
+                {answer.member
+                  ? getMemberDisplayName(answer.member as Member)
+                  : DEFAULT_TEXT_HOLDER}
+              </span>
+            </div>
+            <button
+              onClick={handleLikeClick}
+              disabled={!memberId || !accessToken || isPending}
+              className="flex w-14 cursor-pointer items-center gap-1 disabled:cursor-not-allowed"
+            >
+              {isLoggedIn && hasLiked && (
+                <StarIcon className="text-yellow-400" />
+              )}
+              {isLoggedIn && !hasLiked && <StarIconUnfilled />}
+              {!isLoggedIn && <StarIcon />}
+              <span className="prose-p2-medium text-neutral-600">
+                {getDisplayLikesCount(answer.likesCount)}
+              </span>
+            </button>
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -578,6 +578,134 @@ export const extendGraphqlSchema = graphql.extend(() => {
+                severity: 'WARNING',
+                message: errorMessage,
+                context: {
+                  function: 'getMemberEssayAnswersHasLike',
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/latest-answers/answer-card.tsx:
```diff
@@ -1,45 +1,52 @@
 }: AnswerCardProps) {
-  const displayLikesCount = likesCount > 99 ? '99+' : likesCount.toString()
-
   return (
-    <div className="flex min-h-[182px] w-75 flex-col justify-between rounded-2xl border-2 border-neutral-200 bg-white p-5 tablet:w-full">
+    <div
+      className="flex min-h-[182px] w-75 cursor-pointer flex-col justify-between rounded-2xl border-2 border-neutral-200 bg-white p-5 transition-colors hover:border-neutral-300 tablet:w-full"
+      onClick={onClick}
+    >
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/cms/schema.graphql:
```diff
@@ -2486,6 +2532,13 @@ type Query {
+type getMemberEssayAnswersHasLiked {
+  answerId: ID
+  hasLiked: Boolean
+  likeId: ID
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/all-answers/index.tsx:
```diff
@@ -0,0 +1,217 @@
+        const postId = post.id || post.slug
+        if (!postId) {
+          throw new Error(
+            'Post is missing both id and slug. Data integrity issue detected.'
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/all-answers/post-answer-card/question-answer-card.tsx:
```diff
@@ -0,0 +1,94 @@
+        {hasMoreAnswers && (
+          <button
+            onClick={() => onOpenModal(postSlug)}
+            className="mt-2 flex w-full cursor-pointer items-center justify-center px-5 py-1 prose-p1 text-neutral-600 hover:text-red-400 active:text-red-500"
```
[fix(frontend): copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/b1b5f631f968d442c350a1a2477a51631f6acdb0)

ericsen-tsai commented on packages/cms/lists/member.ts:
```diff
@@ -125,13 +144,12 @@ export default list<ListType<'Member'>>({
       update: operationAccessControl,
       delete: operationAccessControl,
     },
     filter: {
```
Since the `idea-hub` page is visible for all users, some fields should be public to show the correct information.
I've configured other fields as private.

ericsen-tsai commented on packages/frontend/src/components/dialog.tsx:
```diff
@@ -0,0 +1,154 @@
+          'tablet:h-[calc(100vh-112px)] desktop:w-[calc(100vw-96px)]',
+          // HD: max width 1200px, h-[calc(100vh-112px)] (56px padding top/bottom)
+          'tablet:h-[calc(100vh-112px)] hd:max-w-[1200px]',
+          // Remove all focus/active borders inside dialog
+          '[&_*]:focus:!ring-0 [&_*]:focus:!outline-none [&_*]:focus-visible:!ring-0 [&_*]:focus-visible:!outline-none [&_*]:active:!ring-0 [&_*]:active:!outline-none',
```
This is the default style for dialog

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/post-essay-questions-modal/modal-question-card.tsx:
```diff
@@ -0,0 +1,227 @@
+    return map
+  }, [hasLikedData])
+
+  const answerOrderBy: PostEssayAnswerOrderByInput[] = [{ createdAt: 'desc' }]
```
Since there might be other variables added in, I'll leave it as it is

ericsen-tsai commented on packages/frontend/src/modules/idea-hub/utils.ts:
```diff
@@ -0,0 +1,14 @@
+  return member.nickname || member.name || member.email || DEFAULT_TEXT_HOLDER
+}
+
+export const getDisplayLikesCount = (likesCount: number | null | undefined) => {
+  if (!likesCount) return '0'
+  if (likesCount > 99) return '99+'
+  return likesCount > 0 ? likesCount.toString() : '0'
+}
```
This function is to prevent someone from passing a negative number and let the viewer have a "normal" feedback

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -578,6 +580,137 @@ export const extendGraphqlSchema = graphql.extend(() => {
+              return {
+                answerId: id,
+                hasLiked: !!answer,
+                likeId: answer?.id?.toString() ?? '',
```
question:
`essayAnswers`, `answer` 的變數命名有點模糊，這邊指的應該是 `essayAnswerLikes` 和 `essayAnswerLikeId` 嗎？

ericsen-tsai replied:
Fixed in [refactor(cms,frontend): rename answerId and answerLikeId to essayAnswerId and essayAnswerLikeId](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/d5529a0144a52c3318433fcf241f90fc9b705041)

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -578,6 +580,137 @@ export const extendGraphqlSchema = graphql.extend(() => {
+          } catch (err) {
+            let errorMessage = 'getMemberEssayAnswersHasLike failed'
+            if (err instanceof AxiosError) {
+              const annotatedErr = _errors.helpers.annotateAxiosError(err)
```
Note:
`annotateAxiosError` 會把 axios config 印出來，裡面包含 `headers` 和 `auth`，其資訊包含 authorization/session token，要需要 patch 一個新版，遮蔽機敏資訊。

ericsen-tsai replied:
Sure, and I think adding type support for the package is also needed

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -578,6 +580,137 @@ export const extendGraphqlSchema = graphql.extend(() => {
+            console.log(
+              JSON.stringify({
+                severity: 'ERROR',
+                message: errorMessage,
```
suggestion:
如果不是 AxiosError，目前 errorMessage 會印出 'getMemberEssayAnswersHasLike failed'，而這個 message 不會觸發 error reporting。

是否要改成類似 `err instanceof Error ? err.stack : 'getMemberEssayAnswersHasLike failed'` 的寫法？

或是透過 `errors.helpers.wrap(err)` 來處理？

ericsen-tsai replied:
Fixed in [fix(cms): clarify error messages in GCP structured logging](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/34c0f0f76376cc1255c016646ba3d0cf9a5410bf), thanks

nickhsine commented on packages/frontend/src/modules/idea-hub/post-essay-questions-modal/hooks/use-optimistic-like-answer.ts:
```diff
@@ -0,0 +1,227 @@
+        'all-posts',
+        answerOrderBy,
+        answerTake,
+      ]
```
question:
`allPostEssayAnswersQueryKey` 是透過陣列組出 queryKey，
而 `src/api-utils/react-query/hooks/post-essay-answer.ts` 中的 `useAllPostEssayAnswersQuery` function 是透過
```
    queryKey: [
      POST_ESSAY_ANSWERS_QUERY_KEY,
      'all-members',
      'all-posts',
      ...(orderBy ? [JSON.stringify(orderBy)] : []),
      ...(take ? [take] : []),
    ],
```

不確定兩個組出來的 queryKey 是否一致？

ericsen-tsai replied:
You got the point.
Query keys were scattered in many places, I use function to centralize it

[refactor(frontend): use query keys function to get query keys](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/d5312fc765ebe1d311d0f00498719dc98c031e7b)

nickhsine commented on packages/frontend/src/modules/idea-hub/post-essay-questions-modal/modal-question-card.tsx:
```diff
@@ -0,0 +1,229 @@
+  )
+}
+
+export default ModalQuestionCard
```
question:
ModalQuestionCard 會在 mount 時就對每個問題打 infinite query（再加上 hasLiked query），且 defaultAccordionValue 預設全部開啟。若一篇文章問題數多，首次進入 modal 是否會同時觸發 N 個分頁查詢與 N 個 like 查詢？

ericsen-tsai replied:
Do you mean the count of answers?

The modal will trigger 1 (get post questions) + 3 (get question answers) + 3 (get likes for each answers) at the first render.
I think the current scenario is acceptable. If there are more than that scale of questions, I can move the `answer-like` query outside of the answer item and just query once.

nickhsine commented on packages/frontend/src/modules/idea-hub/all-answers/index.tsx:
```diff
@@ -15,12 +16,18 @@ import PostAnswerCardSkeleton from './post-answer-card-skeleton'
 
   const { data, isLoading, isFetchingNextPage, hasNextPage, fetchNextPage } =
-    useGetPostsEssayAnswersWithLikesInfinityQuery({
+    usePostsEssayAnswersWithLikesInfinityQuery({
```
nitpick:

不確定是否可以在 `usePostsEssayAnswersWithLikesInfinityQuery` 新增 `select: transformPostsWithAnswers` property，移動 `allPosts` 的建立邏輯至 `transformPostsWithAnswers` function，讓 React Query 來 transform `data.pages`。

Ex:

```
const transformPostsWithAnswers = () => {
  if (!data?.pages) return []
  const allPosts: PostWithTwoTopLikesAnswersPerQuestion['posts'] = []
  data.pages.forEach((page) => {
     ...
  }
}
usePostsEssayAnswersWithLikesInfinityQuery({
  orderBy: ..., 
  take: ...,
  answerOrderBy: ..,
  select: transformtransformPostsWithAnswers,
})
```

如此一來，在 `data` 沒有改變的情況下，re-render 時就不需要用 `useMemo` 來重跑 `allPosts` 的建立。

ericsen-tsai replied:
Adjusted in [refactor(frontend): enhance essay answers fetching with transformation utility](https://github.com/kids-reporter/kids-reporter-monorepo/pull/863/changes/4a7bab44eb9768898e83a5c0d282e1a1933f7017)

ericsen-tsai committed:
refactor(frontend): use query keys function to get query keys

ericsen-tsai committed:
refactor(frontend): enhance essay answers fetching with transformation utility

- Added a new utility function to transform infinite query data for posts with essay answers and likes.
- Updated the `usePostsEssayAnswersWithLikesInfinityQuery` hook to include a `select` option for data transformation.
- Refactored the AllAnswers component to utilize the new transformation utility, improving data handling and rendering logic.
- Adjusted the default page item count for better pagination control.

ericsen-tsai committed:
refactor(cms,frontend): rename answerId and answerLikeId to essayAnswerId and essayAnswerLikeId

ericsen-tsai committed:
fix(cms): clarify error messages in GCP structured logging

nickhsine commented:
LGTM 👍 👍 👍

ericsen-tsai committed:
fix(cms): rename symbol
